cardano-cli/src/Cardano/CLI/Shelley/Run.hs

```diff
@@ -86,28 +86,57 @@ runBlockCmd cmd = liftIO $ putStrLn $ "TODO: runBlockCmd: " ++ show cmd
 runSystemCmd:: SystemCmd -> ExceptT ShelleyClientCmdError IO ()
 runSystemCmd cmd = liftIO $ putStrLn $ "TODO: runSystemCmd: " ++ show cmd
 
-runGovernanceCmd :: GovernanceCmd -> ExceptT ShelleyClientCmdError IO ()
-runGovernanceCmd (GovernanceUpdateProposal out eNo genVKeys ppUp) =
-  firstExceptT ShelleyCmdGovernanceError $ runGovernanceUpdateProposal out eNo genVKeys ppUp
+runGovernanceCmd :: GovernanceCmd -> ExceptT ShelleyGovernanceError IO ()
+runGovernanceCmd (GovernanceMIRCertificate vKeys rewards out) = runGovernanceMIRCertificate vKeys rewards out
+runGovernanceCmd (GovernanceUpdateProposal out eNo genVKeys ppUp) = runGovernanceUpdateProposal out eNo genVKeys ppUp
 runGovernanceCmd cmd = liftIO $ putStrLn $ "TODO: runGovernanceCmd: " ++ show cmd
 
 data ShelleyGovernanceError
-  = GovernanceWriteUpdateError !FilePath !ApiError
+  = GovernanceWriteMIRCertError !FilePath !ApiError
+  | GovernanceWriteUpdateError !FilePath !ApiError
   | GovernanceEmptyUpdateProposal
   | GovernanceReadGenVerKeyError !FilePath !KeyError
+  | GovernanceReadStakeVerKeyError !FilePath !ApiError
   deriving Show
 
 renderShelleyGovernanceError :: ShelleyGovernanceError -> Text
 renderShelleyGovernanceError err =
   case err of
     GovernanceReadGenVerKeyError fp keyErr ->
       "Error reading genesis verification key at: " <> textShow fp <> " Error: " <> renderKeyError keyErr
+    GovernanceReadStakeVerKeyError stKeyFp apiErr ->
+      "Error reading stake key at: " <> textShow stKeyFp <> " Error: " <> renderApiError apiErr
+    GovernanceWriteMIRCertError certFp apiErr ->
+      "Error writing MIR certificate at: " <> textShow certFp <> " Error: " <> renderApiError apiErr
     GovernanceWriteUpdateError fp apiError ->
       "Error writing shelley update proposal at: " <> textShow fp <> " Error: " <> renderApiError apiError
     -- TODO: The equality check is still not working for empty update proposals.
     GovernanceEmptyUpdateProposal ->
       "Empty update proposals are not allowed"
 
+runGovernanceMIRCertificate
+  :: [VerificationKeyFile]
+  -- ^ Stake verification keys
+  -> [ShelleyCoin]
+  -- ^ Reward amounts
+  -> OutputFile
+  -> ExceptT ShelleyGovernanceError IO ()
+runGovernanceMIRCertificate vKeys rwdAmts (OutputFile oFp) = do
+    sCreds <- mapM readStakeKey vKeys
+
+    let mirCert = shelleyMIRCertificate $ mirMap sCreds rwdAmts
```

Comment:
We should check that the number of `vKeys` and `rwdAmts` are the same and fail with an appropriate error if not. Currently we use `zip` below which means we ignore the longer ones from either side.